src/datasources/asset/AssetDataSource.ts

```diff
@@ -0,0 +1,106 @@
+import { DataSourceBase } from 'core/DataSourceBase';
+import {
+  AssetFilterProperties,
+  AssetModel, AssetQuery,
+  AssetQueryType,
+  AssetsResponse,
+} from './types';
+import { getWorkspaceName, replaceVariables } from "../../core/utils";
+import { SystemMetadata } from "../system/types";
+import { defaultOrderBy, defaultProjection } from "../system/constants";
+
+export class AssetDataSource extends DataSourceBase<AssetQuery> {
+  constructor(
+    readonly instanceSettings: DataSourceInstanceSettings,
+    readonly backendSrv: BackendSrv = getBackendSrv(),
+    readonly templateSrv: TemplateSrv = getTemplateSrv()
+  ) {
+    super(instanceSettings, backendSrv, templateSrv);
+  }
+
+  baseUrl = this.instanceSettings.url + '/niapm/v1';
+
+  defaultQuery = {
+    type: AssetQueryType.Metadata,
+    workspace: '',
+    minionIds: [],
+  };
+
+  async runQuery(query: AssetQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
+    return await this.processMetadataQuery(query)
+  }
+
+  async processMetadataQuery(query: AssetQuery) {
+    const result: DataFrameDTO = { refId: query.refId, fields: [] };
+    const minionIds = replaceVariables(query.minionIds, this.templateSrv);
+    let workspaceId = this.templateSrv.replace(query.workspace);
+    const conditions = [];
+    if (workspaceId) {
+      conditions.push(`workspace = "${workspaceId}"`);
+    }
+    if (minionIds.length) {
+      const systemsCondition = minionIds.map(id => `${AssetFilterProperties.LocationMinionId} = "${id}"`)
+      conditions.push(`(${systemsCondition.join(' or ')})`);
+    }
+    const assetFilter = conditions.join(' and ');
+    const assets: AssetModel[] = await this.queryAssets(assetFilter);
+    const workspaces = await this.getWorkspaces();
+    result.fields = [
+      { name: 'id', values: assets.map(a => a.id) },
+      { name: 'name', values: assets.map(a => a.name) },
+      { name: 'model name', values: assets.map(a => a.modelName) },
+      { name: 'serial number', values: assets.map(a => a.serialNumber) },
+      { name: 'bus type', values: assets.map(a => a.busType) },
+      { name: 'asset type', values: assets.map(a => a.assetType) },
+      { name: 'is NI asset', values: assets.map(a => a.isNIAsset) },
+      { name: 'calibration status', values: assets.map(a => a.calibrationStatus) },
+      { name: 'is system controller', values: assets.map(a => a.isSystemController) },
+      { name: 'last updated timestamp', values: assets.map(a => a.lastUpdatedTimestamp) },
+      { name: 'minionId', values: assets.map(a => a.location.minionId) },
+      { name: 'parent name', values: assets.map(a => a.location.parent) },
+      { name: 'workspace', values: assets.map(a => getWorkspaceName(workspaces, a.workspace)) },
+    ];
+    return result;
+  }
+
+
+  shouldRunQuery(_: AssetQuery): boolean {
+    return true;
+  }
+
+  async queryAssets(filter = '', take = -1): Promise<AssetModel[]> {
```

Comment:
take=-1 might not work when there are many assets (e.g. 100k).
One possible solution is to get data in batches:
- Request1: skip 0, take 5000
- Request2: skip 5000, take 5000

I'm not saying we need to handle this now but might worth having a work item to address this in the future.